README.md
# extra-config-json

[![Marketplace Version](https://vsmarketplacebadge.apphb.com/version/jawa0919.extra-config-json.svg)](https://marketplace.visualstudio.com/items?itemName=jawa0919.extra-config-json) [![Installs](https://vsmarketplacebadge.apphb.com/installs/jawa0919.extra-config-json.svg)](https://marketplace.visualstudio.com/items?itemName=jawa0919.extra-config-json) [![Rating](https://vsmarketplacebadge.apphb.com/rating-star/jawa0919.extra-config-json.svg)](https://marketplace.visualstudio.com/items?itemName=jawa0919.extra-config-json) [![GitHub issues](https://img.shields.io/github/issues/jawa0919/extra-config-json)](https://github.com/jawa0919/extra-config-json/issues)![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/jawa0919/extra-config-json)

为项目配置文件 `extraConfig.json`/`.env.*.local` 开发的 vscode 插件

快速切换，简单保存，忽略 Git, 生成.env

Quick swap, Simple Save, Git Unchanged and Generate .env

# 功能

- [x] 备份 `extraConfig.json` 到用户目录文件夹(可配置)
- [x] 快速替换本地备份保存的 `extraConfig.json`
- [x] 切换 `extraConfig.json` 自动本地记录历史文件，方便查找溯源
- [x] 忽略 `extraConfig.json` 的变更，`git`提交时无视变更
- [x] `extraConfig.json` 生成 `.env.*.local`
- [x] `.env.*` 生成 `extraConfig.json`
- [x] 生成 `.env.*.local` 时，筛选不符合`vue`的字段前缀 `VUE_APP_` 的数据(可配置)
- [x] 生成 `.env.*.local` 时，筛选不符合`react`的字段前缀 `REACT_APP_` 的数据(可配置)
- [x] 生成 `.env.*.local` 时，筛选不符合`vite`的字段前缀 `VITE_` 的数据(可配置)
- [x] 生成 `.env.*.local` 时，在 `package.json` 中添加执行脚本
- [x] 文件互相转换时，优化对象的层级对 json 文件的影响

---

# 文件互转说明

1. json2env 时字段丢失。 Vue/React 部分字段无法通过`process.env.`编译。

2. json2env 时字段丢失。支持`vite`的字段前缀 `VITE_`。

3. env 文件是使用单行的键值对 k/v 系统存储数据，json 字段可以嵌套对象的情况，二级字段使用`JSON.stringify` 将对象转换为字符串，开发时请使用`JSON.parse` 将字符串转换为对象

   ```json
   {
     "VUE_APP_A": "qwe",
     "VUE_APP_B": 10,
     "VUE_APP_C": 100.1,
     "VUE_APP_D": true,
     "VUE_APP_E": ["asd", 200, 200.2, true],
     "VUE_APP_F": {
       "a": "zxc",
       "b": 300,
       "c": 300.003,
       "d": true,
       "e": [400, 400.2, true],
       "f": {
         "h": ["g", 500, 0.00005, false]
       }
     }
   }
   ```

   ```env
   # .env.dev.local
   VUE_APP_A = qwe
   VUE_APP_B = 10
   VUE_APP_C = 100.1
   VUE_APP_D = true
   VUE_APP_E = ["asd",200,200.2,true]
   VUE_APP_F = {"a":"zxc","b":300,"c":300.003,"d":true,"e":[400,400.2,true],"f":{"h":["g",500,0.00005,false]}}
   ```

   ![env_vk](./docs/img/env_vk.png)

## 切换/保存

在 `extraConfig.json` 文件右上方有两个功能按钮，点击可以切换/保存`extraConfig.json`文件

![swap](./docs/img/swap.gif)

## 记录历史

切换`extraConfig.json`文件时会备份文件夹下的`History`文件夹中备份

![history](./docs/img/history.png)

## git 假设提交

使用`git`假设提交命令，忽略`extraConfig.json`的变更,只对本机生效

右键点击`extraConfig.json`文件时出现两个文件功能选项

- `Git --assume-unchanged @extra-config-json`

  点击后本地修改`extraConfig.json`时，不会触发`git`变更记录，保留其初始版本

- `Git --no-assume-unchanged @extra-config-json` ，

  点击后取消不触发状态

---

![git_assume](./docs/img/git_assume.png)

```shell
# ignore一个文件的更改又保留其初始版本
git update-index --assume-unchanged [file-path]
# 取消
git update-index --no-assume-unchanged [file-path]
```

## env 互相转换 json

右键点击`extraConfig.json`/`.env.*.local` 文件时出现相关功能选项

![env_to_json](./docs/img/env_to_json.png)

## 配置

- 默认备份文件夹为`用户文件目录`下的`.extra-config-json`文件夹

  也可自行配置，在 `vscode`的设置中找到如下配置项，写入文件夹地址

- 默认对 json 文件筛选 `VUE_APP_XXX`开头的字段

- 默认对 json 文件筛选 `REACT_APP_XXX`开头的字段

---

![setting_all](./docs/img/setting_all.png)

## 其他

![home](./docs/img/home.png)

欢迎大家 `提出想法` 和 `反馈问题` [issues](https://github.com/jawa0919/extra-config-json/issues)


CHANGELOG.md
# Change Log

# 0.1.4

- fix ex

# 0.1.3

- add Matches

# 0.1.2

- add vite field prefix `VITE_`

# 0.1.1

- fix env2json bug

# 0.1.0

- fix auto add pakcage.json script

# 0.0.4

- add json2env/env2json

# 0.0.3

- fix readme img

# 0.0.2

- fix save bug
- fix readme

# 0.0.1

- Initial release


src/core/util.ts
/*
 * @FilePath     : /src/core/util.ts
 * @Date         : 2022-04-30 21:51:35
 * @Author       : jawa0919 <jawa0919@163.com>
 * @Description  : 工具
 */

import { TextDocument, window, workspace, WorkspaceFolder } from "vscode";

export function findTextDocument(fsPath: string): TextDocument | undefined {
  return workspace.textDocuments.find((e) => fsPath === e.uri.fsPath);
}

export function findWorkspaceFolder(fsPath: string): WorkspaceFolder | undefined {
  return workspace.workspaceFolders?.find((r) => fsPath.startsWith(r.uri.fsPath));
}

export async function showInputBox(): Promise<string | undefined> {
  let opt = {
    title: "Please Input File Name",
    placeHolder: "Please Input File Name eg: gat sdk gz_dev py_test",
  };
  return await window.showInputBox(opt);
}

export async function showQuickPick(fileList: string[]): Promise<string | undefined> {
  return await window.showQuickPick(fileList);
}

export async function showInformationMessage(message: string): Promise<string | undefined> {
  return await window.showInformationMessage(message);
}

export async function showErrorMessage(message: string): Promise<string | undefined> {
  return await window.showErrorMessage(message);
}

export function dateTimeName(): string {
  return new Date().toISOString().substring(0, 19).replace(/[-:T]/g, "");
}

export function object2Kv(res: Record<string, any>): Record<string, string> {
  const temp: Record<string, string> = {};
  Object.keys(res).forEach((key) => {
    if (typeof res[key] === "object") {
      temp[key] = JSON.stringify(res[key]);
    } else {
      temp[key] = `${res[key]}`;
    }
  });
  return temp;
}

export function kv2Object(res: Record<string, string>): Record<string, any> {
  const temp: Record<string, any> = {};
  Object.keys(res).forEach((key) => {
    const value = res[key];
    try {
      temp[key] = JSON.parse(value);
    } catch (error) {
      if (value === "true" || value === "false") {
        temp[key] = Boolean(value);
      } else if (Number(value)) {
        temp[key] = Number(value);
      } else {
        temp[key] = value;
      }
    }
  });
  return temp;
}


src/core/config.ts
/*
 * @FilePath     : /src/core/config.ts
 * @Date         : 2022-04-30 20:56:00
 * @Author       : jawa0919 <jawa0919@163.com>
 * @Description  : 配置文件
 */

import { homedir } from "os";
import { join } from "path";
import { workspace } from "vscode";

export const name = "extra-config-json";

export const historyDirName = ".history";

export let dirPath = join(homedir(), `.${name}`);
export let onlyViteField = true;
export let onlyVueAppField = true;
export let onlyReactAppField = true;

export const envLocalNameList = [
  ".env.local",
  ".env.dev.local",
  ".env.development.local",
  ".env.test.local",
  ".env.qa.local",
  ".env.uat.local",
  ".env.stag.local",
  ".env.staging.local",
  ".env.production.local",
];

export function initConfig(): void {
  let cf = workspace.getConfiguration();
  let path = cf.get<string>("extra-config-json.dirPath") || dirPath;
  dirPath = path;
  onlyViteField = cf.get<boolean>("extra-config-json.onlyViteField") || onlyViteField;
  onlyVueAppField = cf.get<boolean>("extra-config-json.onlyVueAppField") || onlyVueAppField;
  onlyReactAppField = cf.get<boolean>("extra-config-json.onlyReactAppField") || onlyReactAppField;
}


src/core/envFile.ts
/*
 * @FilePath     : /src/core/envFile.ts
 * @Date         : 2022-04-30 20:37:38
 * @Author       : jawa0919 <jawa0919@163.com>
 * @Description  : env文件处理
 */

import { ensureFileSync, existsSync, readFileSync, writeFileSync } from "fs-extra";
import { join } from "path";
import { commands } from "vscode";
import { saveSwapHistory } from "./jsonFile";
import { findWorkspaceFolder, kv2Object } from "./util";

export const envFileCMD = [
  commands.registerCommand("extra-config-json.env.save", _envSave),
  commands.registerCommand("extra-config-json.env.to.json", _envToJson),
];

async function _envSave(...res: any[]) {
  console.log("_envSave");
  let fsPath: string = res[0].fsPath || "";
  if (fsPath === "") return;
}

async function _envToJson(...res: any[]) {
  console.log("_envToJson");
  let fsPath: string = res[0].fsPath || "";
  if (fsPath === "") return;

  let envKvData: Record<string, string> = {};
  readFileSync(fsPath, "utf-8")
    .split("\n")
    .forEach((line) => {
      if (line.startsWith("#")) return;
      let [key, value] = line.split("=").map((v) => v.trim());
      envKvData[key] = value;
    });
  let jsonObjectData = kv2Object(envKvData);

  let wf = findWorkspaceFolder(fsPath);
  let jsonPath = join(wf?.uri.fsPath || "", "public", `extraConfig.json`);
  if (existsSync(jsonPath)) saveSwapHistory(jsonPath);
  ensureFileSync(jsonPath);
  writeFileSync(jsonPath, JSON.stringify(jsonObjectData, null, 2), "utf-8");
}


src/core/pkgFile.ts
/*
 * @FilePath     : /src/core/pkgFile.ts
 * @Date         : 2022-04-30 20:38:13
 * @Author       : jawa0919 <jawa0919@163.com>
 * @Description  : package.json文件处理
 */

import { readJsonSync, writeJsonSync } from "fs-extra";
import { join } from "path";
import { onlyVueAppField, onlyReactAppField, onlyViteField } from "./config";
import { findWorkspaceFolder } from "./util";

interface PkgData {
  scripts?: {
    serve?: string;
    start?: string;
    dev?: string;
    [key: string]: any;
  };
  dependencies?: {
    vue?: string;
    react?: string;
    [key: string]: any;
  };
  devDependencies?: {
    vite?: string;
    [key: string]: any;
  };
  [key: string]: any;
}

export function loadPkgData(fsPath: string): PkgData {
  console.log("loadPkgData");
  let wf = findWorkspaceFolder(fsPath);

  let pkgPath = join(wf?.uri.fsPath || "", "package.json");
  // let pkgData = await require("read-pkg-up")({ cwd: pkgPath });
  let pkgData: PkgData = readJsonSync(pkgPath);
  return pkgData;
}

export function filterField(jsonData: Record<string, string>, pkgData: PkgData): Record<string, string> {
  console.log("filterField");
  if (pkgData.devDependencies?.vite) {
    return viteField(jsonData, pkgData);
  } else if (pkgData.dependencies?.vue) {
    return vueField(jsonData, pkgData);
  } else if (pkgData.dependencies?.react) {
    return reactField(jsonData, pkgData);
  } else {
    return jsonData;
  }
}

export function viteField(jsonData: Record<string, string>, pkgData: PkgData): Record<string, string> {
  console.log("viteField");
  if (onlyViteField) {
    let temp: Record<string, string> = {};
    Object.keys(jsonData).forEach((e) => {
      if (e.startsWith("VITE_")) {
        temp[e] = jsonData[e];
      }
    });
    return temp;
  }
  return jsonData;
}

export function vueField(jsonData: Record<string, string>, pkgData: PkgData): Record<string, string> {
  console.log("vueField");
  if (onlyVueAppField) {
    let temp: Record<string, string> = {};
    Object.keys(jsonData).forEach((e) => {
      if (e.startsWith("VUE_APP_")) {
        temp[e] = jsonData[e];
      }
    });
    return temp;
  }
  return jsonData;
}

export function reactField(jsonData: Record<string, string>, pkgData: PkgData): Record<string, string> {
  console.log("reactField");
  if (onlyReactAppField) {
    let temp: Record<string, string> = {};
    Object.keys(jsonData).forEach((e) => {
      if (e.startsWith("REACT_APP_")) {
        temp[e] = jsonData[e];
      }
    });
    return temp;
  }
  return jsonData;
}

export function addScriptInPkg(envLocalName: string, pkgData: PkgData, fsPath: string) {
  console.log("addScriptInPkg");
  let mode = envLocalName.replace(/.env./, "");

  let cmdSuffix = `--mode ${mode}`;
  pkgData.scripts = pkgData.scripts || {};
  if (pkgData.scripts.serve) {
    let name = `serve:${mode}`;
    if (pkgData.scripts?.[name]) return;
    // FIXME 2022-05-01 11:55:40 serve vue-cli
    let cmd = pkgData.scripts.serve;
    if (cmd.includes(cmdSuffix)) return;
    cmd += ` ${cmdSuffix}`;
    pkgData.scripts[name] = cmd;
  } else if (pkgData.scripts.start) {
    let name = `start:${mode}`;
    if (pkgData.scripts?.[name]) return;
    // FIXME 2022-05-01 11:55:40 start react-create-app
    let cmd = pkgData.scripts.start;
    if (cmd.includes(cmdSuffix)) return;
    cmd += ` ${cmdSuffix}`;
    pkgData.scripts[name] = cmd;
  } else if (pkgData.scripts.dev) {
    let name = `dev:${mode}`;
    if (pkgData.scripts?.[name]) return;
    // FIXME 2022-05-01 11:55:40 dev vite
    let cmd = pkgData.scripts.dev;
    if (cmd.includes(cmdSuffix)) return;
    cmd += ` ${cmdSuffix}`;
    pkgData.scripts[name] = cmd;
  } else if (pkgData.dependencies?.vue) {
    let name = `serve:${mode}`;
    if (pkgData.scripts?.[name]) return;
    pkgData.scripts[name] = `vue-cli-service serve ${cmdSuffix}`;
  } else if (pkgData.dependencies?.react) {
    let name = `start:${mode}`;
    if (pkgData.scripts?.[name]) return;
    pkgData.scripts[name] = `react-scripts start ${cmdSuffix}`;
  }

  let wf = findWorkspaceFolder(fsPath);
  let pkgPath = join(wf?.uri.fsPath || "", "package.json");

  writeJsonSync(pkgPath, pkgData, { spaces: 2 });
}


src/core/jsonFile.ts
/*
 * @FilePath     : /src/core/jsonFile.ts
 * @Date         : 2022-04-30 20:37:09
 * @Author       : jawa0919 <jawa0919@163.com>
 * @Description  : json文件处理
 */

import { execFileSync } from "child_process";
import { copyFileSync, ensureDirSync, readdirSync, readJsonSync, writeFileSync } from "fs-extra";
import { join } from "path";
import { commands } from "vscode";
import { dirPath, envLocalNameList, historyDirName } from "./config";
import { addScriptInPkg, loadPkgData, filterField } from "./pkgFile";
import {
  findTextDocument,
  showErrorMessage,
  showInformationMessage,
  showInputBox,
  showQuickPick,
  dateTimeName,
  findWorkspaceFolder,
  object2Kv,
} from "./util";

export const jsonFileCMD = [
  commands.registerCommand("extra-config-json.json.save", _jsonSave),
  commands.registerCommand("extra-config-json.json.swap", _jsonSwap),
  commands.registerCommand("extra-config-json.json.to.env", _jsonToEnv),
  commands.registerCommand("extra-config-json.json.git.assume", _jsonGitAssume),
  commands.registerCommand("extra-config-json.json.git.no.assume", _jsonGitNoAssume),
];

async function _jsonSave(...res: any[]) {
  console.log("_jsonSave");
  let fsPath: string = res[0].fsPath || "";
  if (fsPath === "") return;

  let td = findTextDocument(fsPath);
  await td?.save();

  let name = await showInputBox();
  if (name) {
    ensureDirSync(dirPath);
    let jsonPath = join(dirPath, `${name}.json`);
    copyFileSync(fsPath, jsonPath);
  }
}

async function _jsonSwap(...res: any[]) {
  console.log("_jsonSwap");
  let fsPath: string = res[0].fsPath || "";
  if (fsPath === "") return;

  let td = findTextDocument(fsPath);
  if (td?.isDirty) {
    showErrorMessage("Please Save File Before");
    return;
  }

  ensureDirSync(dirPath);
  let jsonFileNameList = readdirSync(dirPath).filter((e) => e.endsWith(".json"));
  if (jsonFileNameList.length === 0) {
    showInformationMessage("No Find Json");
    return;
  }

  let jsonFileName = await showQuickPick(jsonFileNameList);
  if (jsonFileName) {
    saveSwapHistory(fsPath);
    ensureDirSync(dirPath);
    let jsonFilePath = join(dirPath, jsonFileName);
    copyFileSync(jsonFilePath, fsPath);
  }
}

async function _jsonToEnv(...res: any[]) {
  console.log("_jsonToEnv");
  let fsPath: string = res[0].fsPath || "";
  if (fsPath === "") return;

  let envLocalName = await showQuickPick(envLocalNameList);
  if (envLocalName) {
    const jsonObjectData: Record<string, any> = readJsonSync(fsPath);

    let envKvData = object2Kv(jsonObjectData);
    let pkgData = loadPkgData(fsPath);

    envKvData = filterField(envKvData, pkgData);

    let wf = findWorkspaceFolder(fsPath);
    let bf = [`# ${envLocalName}`];
    Object.keys(envKvData).forEach((k) => bf.push(`${k} = ${envKvData[k]}`));
    let envPath = join(wf?.uri.fsPath || "", envLocalName);
    writeFileSync(envPath, bf.join("\n"), "utf-8");
    addScriptInPkg(envLocalName, pkgData, fsPath);
  }
}

async function _jsonGitAssume(...res: any[]) {
  console.log("_jsonGitAssume");
  let fsPath: string = res[0].fsPath || "";
  if (fsPath === "") return;

  // let cmd = "git update-index --assume-unchanged " + fsPath;
  let wf = findWorkspaceFolder(fsPath);
  let cwd = wf?.uri.fsPath;
  execFileSync("git", ["update-index", "--assume-unchanged", fsPath], { cwd });
}

async function _jsonGitNoAssume(...res: any[]) {
  console.log("_jsonGitNoAssume");
  let fsPath: string = res[0].fsPath || "";
  if (fsPath === "") return;

  // let cmd = "git update-index --no-assume-unchanged " + fsPath;
  let wf = findWorkspaceFolder(fsPath);
  let cwd = wf?.uri.fsPath;
  execFileSync("git", ["update-index", "--no-assume-unchanged", fsPath], { cwd });
}

////////////////////////////////////////////////////////////////////////////////

export async function saveSwapHistory(fsPath: string) {
  console.log("saveSwapHistory");
  if (fsPath === "") return;

  let name = dateTimeName();
  ensureDirSync(dirPath);
  let historyDirPath = join(dirPath, historyDirName);
  ensureDirSync(historyDirPath);

  let historyFilePath = join(historyDirPath, `${name}.json`);
  copyFileSync(fsPath, historyFilePath);
}
